src/eval.ml

```diff
@@ -218,11 +263,82 @@ and eq_modulo : term -> term -> bool = fun a b ->
   let res = try eq_modulo [(a,b)]; true with Exit -> false in
   if !log_enabled then log_eqmd (r_or_g res "%a == %a") pp a pp b; res
 
+(** [tree_walk t d s] tries to match stack [s] against tree [t] of depth [d]. *)
+and tree_walk : Dtree.t -> int -> term list -> term option = fun tree d stk ->
+  let vars = if d > 0 then Array.make d Kind else [||] in (* dummy terms *)
+  let module R = Dtree.ReductionStack in
+  let stk = R.of_list stk in
+  (* [walk t s c] where [s] is the stack of terms to match and [c] the cursor
+     indicating where to write in the [env] array described in {!module:Terms}
+     as the environment of the RHS during matching. *)
+  let rec walk tree stk cursor =
+    match tree with
+    | Fail                                   -> None
+    | Leaf(env_builder, act)                 ->
+        (* Allocate an environement for the action. *)
+        let env = Array.make (Bindlib.mbinder_arity act) TE_None in
+        (* Retrieve terms needed in the action from the [vars] array. *)
+        let fn pos slot =
+          let t = vars.(pos) in
+          let b = Bindlib.raw_mbinder [||] [||] 0 mkfree (fun _ -> t) in
+          env.(slot) <- TE_Some(b)
+        in
+        IntMap.iter fn env_builder;
```

Comment:
Why not using Array.init?

Reply:
The mapping would have to be reversed, wouldn't it?